src/MassTransit/Pipeline/ConsumerFactories/BatchConsumerFactory.cs

```diff
@@ -65,18 +84,67 @@ void IProbeSite.Probe(ProbeContext context)
 
         async Task<BatchConsumer<TConsumer, TMessage>> Add(ConsumeContext<TMessage> context)
         {
-            if (_currentConsumer != null)
+            var groupKey = GetGroupKey(context);
+            BatchConsumer<TConsumer, TMessage> consumer = _currentConsumers.Get(groupKey);
+
+            if (consumer != null)
             {
                 if (context.GetRetryAttempt() > 0)
-                    await _currentConsumer.ForceComplete().ConfigureAwait(false);
+                    await consumer.ForceComplete().ConfigureAwait(false);
+            }
+
+            if (consumer == null || consumer.IsCompleted)
+            {
+                consumer = new BatchConsumer<TConsumer, TMessage>(_messageLimit, _timeLimit, _collector, _dispatcher, _consumerFactory, _consumerPipe);
+                _currentConsumers.Set(groupKey, consumer);
+            }
+
+            await consumer.Add(context).ConfigureAwait(false);
+
+            return consumer;
+        }
+
+        object GetGroupKey(ConsumeContext context)
+        {
+            return _groupingExpression?.Invoke(context);
+        }
+
+
+        class CurrentConsumersDictionary
```

Comment:
Should be for sure concurrent friendly for this specific use case

Reply:
I think that `_collector = new ChannelExecutor(1)` takes care of this; and for the Release call, I think that concurrency is hanlded by `BatchConsumer.IsCompleted` and the `if` statement in line 96 ... not 100% sure though

Reply:
I meant why we even need `Release` method to be exposed outside, you can use IDisposable for `Set`, so when you end - call .Dispose() instead Release

Reply:
So the real difference in thinking is that the grouping should be separated from the batch, and that makes it more like a partition style (similar to the partition filter) but with a dynamic number of groups that are created as new keys are matched, and then deleted once there are no messages for that group id. That way, it would be separated from the batch logic entirely and each batch would be for a particular group.

It keeps the batch logic simple, and focuses on the grouping separately.

Reply:
What about old friend MT's partitioner?

Reply:
I think I understand the idea but I have no clue as to how to implement that yet :)

Reply:
@phatboyg would you have one instance of BatchConsumerFactory per each group/partition?

Reply:
Yeah, I think so. The KeyFilter in GreenPipes is similar to how I'm thinking about it as well, so that the first key in creates an entry and a counter is maintained for each message that comes in, and once all the messages have exited it removes that key from the filter. I need to think about locking support in that regard though. Again, I'm just looking through other code with similar approaches and thinking how best to support routing of messages by key into their own batch consumers.